qa/tasks/quiescer.py

```diff
@@ -0,0 +1,295 @@
+    """
+    The Quiescer does periodic quiescing of the configured paths, by default - the root '/'.
+
+    quiesce_timeout: [1..),default=30 :: maximum time in seconds to wait for the quiesce to succeed
+    quiesce_factor: [0.005..0.5],default=0.1 :: the fraction of the total runtime we want the system quiesced
+    min_quiesce: [1..),default=10 :: the minimum pause time in seconds
+    max_quiesce: [1..),default=60 :: the maximum pause time in seconds
+    initial_delay: [0..),default=120 :: the time in seconds before the first quiesce
+    seed: default=None :: an optional seed to a pseudorandom sequence of quiesce durations
+    roots: List[String], default=["/"] :: the roots to quiesce
+    cancelations_cap: [-1..),default=10 :: the number of times we ignore canceled quiesce sets
+    split_if_longer: int,default=mean(min_quiesce, max_quiesce) :: if the duration is longer than this,
+                     it will be split into two back-to-back half durations
+    """
+
+    MAX_QUIESCE_FACTOR = 0.5    # 50%
+    MIN_QUIESCE_FACTOR = 0.005  # 0.5%
+    QDB_CMD_TIMEOUT    = 30     # sec
+
+    def __init__(self, fs,
+                 quiesce_timeout=30,
+                 quiesce_factor=0.1,
+                 min_quiesce=10,
+                 max_quiesce=60,
+                 initial_delay=120,
+                 cancelations_cap=10,
+                 seed=None,
+                 roots=None,
+                 split_if_longer=None,
+                 **unused_kwargs):
+        super(Quiescer, self).__init__()
+
+        self.logger = log.getChild('fs.[{f}]'.format(f=fs.name))
+        self.fs = fs
+        self.name = 'quiescer.fs.[{f}]'.format(f=fs.name)
+
+        if seed is None:
+            seed = random.randint(0, 999999) # 1M seems sufficient and avoids possible huge numbers
+        self.logger.info(f"Initializing Quiescer with random seed {seed}")
+        self.rnd = random.Random(seed)
+
+        self.quiesce_timeout = quiesce_timeout
+
+        if (quiesce_factor > self.MAX_QUIESCE_FACTOR):
+            self.logger.warn("Capping the quiesce factor at %f (requested: %f)" % (self.MAX_QUIESCE_FACTOR, quiesce_factor))
+            quiesce_factor = self.MAX_QUIESCE_FACTOR
+
+        if quiesce_factor < self.MIN_QUIESCE_FACTOR:
+            self.logger.warn("Setting the quiesce factor to %f (requested: %f)" % (self.MIN_QUIESCE_FACTOR, quiesce_factor))
+            quiesce_factor = self.MIN_QUIESCE_FACTOR
+
+        self.quiesce_factor = quiesce_factor
+        self.min_quiesce = max(1, min_quiesce)
+        self.max_quiesce = max(1, max_quiesce)
+        self.initial_delay = max(0, initial_delay)
+        self.roots = roots or ["/"]
+        self.cancelations_cap = cancelations_cap
+
+        # this can be used to exercise repeated quiesces with minimal delay between them 
+        self.split_if_longer = split_if_longer if split_if_longer is not None else (self.min_quiesce + self.max_quiesce) / 2
+
+    def next_quiesce_duration(self):
```

Comment:
This method needs an comment about what a duration is and what the computation is.

Reply:
Done